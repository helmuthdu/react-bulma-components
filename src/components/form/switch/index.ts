import './switch.sass';

export * from './switch';
