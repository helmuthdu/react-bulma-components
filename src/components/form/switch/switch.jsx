import cn from 'classnames';
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import CONSTANTS from '../../../constants';
import modifiers from '../../../modifiers';

const colors = Object.values(CONSTANTS.COLORS);

export const Switch = ({ className, style, disabled, value, children, checked, color, size, name, ...allProps }) => {
  const props = modifiers.clean(allProps);
  const [isMouseDown, setMouseDown] = useState(false);
  return (
    <label
      data-testid="switch-label"
      className={cn('switch', modifiers.getClassName(allProps), className)}
      disabled={disabled}
      onMouseDown={() => setMouseDown(true)}
      onMouseOut={() => setMouseDown(false)}
      onMouseUp={() => setMouseDown(false)}
      style={style}
    >
      <input
        {...props}
        data-testid="switch-input"
        checked={checked}
        disabled={disabled}
        name={name}
        type="checkbox"
        value={value}
      />
      <span
        data-testid="switch-check"
        className={cn('check', {
          'is-elastic': isMouseDown && !disabled,
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
      />
      <span className="control-label">{children}</span>
    </label>
  );
};

Switch.propTypes = {
  ...modifiers.propTypes,
  checked: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  color: PropTypes.oneOf(colors),
  disabled: PropTypes.bool,
  name: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  size: PropTypes.oneOf(Object.values(CONSTANTS.SIZES)),
  style: PropTypes.object,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

Switch.defaultProps = {
  ...modifiers.defaultProps,
  checked: false
};
