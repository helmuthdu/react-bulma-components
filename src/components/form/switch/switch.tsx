import cn from 'classnames';
import React, { useState } from 'react';
import { Colors } from '../../../constants';
import modifiers, { Modifiers } from '../../../modifiers';

type SwitchProps = Partial<Modifiers> & {
  checked?: boolean;
  color?: Colors;
  disabled?: boolean;
  name?: string;
  onChange: (...args: any[]) => any;
  size?: any;
  value?: string | number;
};

export const Switch: React.FunctionComponent<SwitchProps> = ({
  className,
  style,
  disabled,
  value,
  children,
  checked,
  color,
  size,
  name,
  ...allProps
}: SwitchProps) => {
  const props = modifiers.clean(allProps);
  const [isMouseDown, setMouseDown] = useState(false);
  return (
    <label
      data-testid="switch-label"
      className={cn('switch', modifiers.getClassName(allProps), className)}
      onMouseDown={() => setMouseDown(true)}
      onMouseOut={() => setMouseDown(false)}
      onMouseUp={() => setMouseDown(false)}
      style={style}
    >
      <input
        {...props}
        data-testid="switch-input"
        checked={checked}
        disabled={disabled}
        name={name}
        type="checkbox"
        value={value}
      />
      <span
        data-testid="switch-check"
        className={cn('check', {
          'is-elastic': isMouseDown && !disabled,
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
      />
      <span className="control-label">{children}</span>
    </label>
  );
};

Switch.defaultProps = {
  ...modifiers.defaultProps,
  checked: false
};
