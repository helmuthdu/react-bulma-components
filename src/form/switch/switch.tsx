import clsx from 'clsx';
import * as React from 'react';
import { Colors, Sizes } from '../../constants';
import modifiers, { ElementModifier } from '../../modifiers';

const { useState } = React;

type SwitchProps = ElementModifier & {
  color?: Colors;
  size?: Sizes;
  value?: string | number;
} & Omit<React.ComponentProps<'input'>, 'color' | 'size' | 'unselectable' | 'value'>;

export const Switch: React.FunctionComponent<SwitchProps> = ({
  checked,
  children,
  className,
  color,
  disabled,
  name,
  size,
  style,
  testId,
  value,
  ...rest
}: SwitchProps) => {
  const props = modifiers.clean(rest);
  const [isMouseDown, setMouseDown] = useState(false);
  return (
    <label
      data-testid={(Array.isArray(testId) ? testId[0] : testId) || undefined}
      className={clsx('switch', modifiers.getClassName(rest), className)}
      onMouseDown={() => setMouseDown(true)}
      onMouseOut={() => setMouseDown(false)}
      onMouseUp={() => setMouseDown(false)}
      style={style}
    >
      <input
        data-testid={Array.isArray(testId) ? testId[2] : undefined}
        checked={checked}
        disabled={disabled}
        name={name}
        type="checkbox"
        value={value}
        {...props}
      />
      <span
        data-testid={Array.isArray(testId) ? testId[1] : undefined}
        className={clsx('check', {
          'is-elastic': isMouseDown && !disabled,
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
      />
      <span className="control-label">{children}</span>
    </label>
  );
};

Switch.defaultProps = {
  ...modifiers.defaultProps,
  checked: false
};
