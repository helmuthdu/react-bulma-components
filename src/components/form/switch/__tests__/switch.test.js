import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Switch } from '../switch';

describe('Switch component', () => {
  let onChange;

  beforeEach(() => {
    onChange = jest.fn();
  });

  it('should render', () => {
    const { asFragment } = render(<Switch onChange={onChange}>Text</Switch>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have info color classname', () => {
    const { asFragment } = render(
      <Switch onChange={onChange} color="info">
        Text
      </Switch>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have medium classname', () => {
    const { asFragment } = render(
      <Switch onChange={onChange} size="medium">
        Text
      </Switch>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be disabled', () => {
    const { asFragment } = render(
      <Switch onChange={onChange} disabled>
        Text
      </Switch>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be elastic', () => {
    const { getByText, getByTestId } = render(<Switch onChange={onChange}>Text</Switch>);
    fireEvent.mouseDown(getByText(/Text/i));
    expect(getByTestId('switch-check')).toHaveClass('is-elastic');
  });

  it('should not be elastic', () => {
    const { getByText, getByTestId } = render(<Switch onChange={onChange}>Text</Switch>);
    fireEvent.mouseOut(getByText(/Text/i));
    expect(getByTestId('switch-check')).not.toHaveClass('is-elastic');
  });

  it('should not be elastic 2', () => {
    const { getByText, getByTestId } = render(<Switch onChange={onChange}>Text</Switch>);
    fireEvent.mouseUp(getByText(/Text/i));
    expect(getByTestId('switch-check')).not.toHaveClass('is-elastic');
  });

  it('should set input checked if checked', () => {
    const { getByTestId } = render(<Switch onChange={onChange} checked />);
    expect(getByTestId('switch-input')).toHaveAttribute('checked', '');
  });

  it('should change value on change event', () => {
    const { getByTestId } = render(<Switch onChange={onChange}>Text</Switch>);
    fireEvent.click(getByTestId('switch-label'));
    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
