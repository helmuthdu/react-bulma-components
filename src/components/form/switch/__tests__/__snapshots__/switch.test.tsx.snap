// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component should be disabled 1`] = `
<DocumentFragment>
  <label
    class="switch"
  >
    <input
      disabled=""
      type="checkbox"
      value=""
    />
    <span
      class="check"
    />
    <span
      class="control-label"
    >
      Text
    </span>
  </label>
</DocumentFragment>
`;

exports[`Switch component should have info color classname 1`] = `
<DocumentFragment>
  <label
    class="switch"
  >
    <input
      type="checkbox"
      value=""
    />
    <span
      class="check is-info"
    />
    <span
      class="control-label"
    >
      Text
    </span>
  </label>
</DocumentFragment>
`;

exports[`Switch component should have medium classname 1`] = `
<DocumentFragment>
  <label
    class="switch"
  >
    <input
      type="checkbox"
      value=""
    />
    <span
      class="check is-medium"
    />
    <span
      class="control-label"
    >
      Text
    </span>
  </label>
</DocumentFragment>
`;

exports[`Switch component should render 1`] = `
<DocumentFragment>
  <label
    class="switch"
  >
    <input
      type="checkbox"
      value=""
    />
    <span
      class="check"
    />
    <span
      class="control-label"
    >
      Text
    </span>
  </label>
</DocumentFragment>
`;
